### PR TITLE
Password reset security fix

### LINKED ISSUE #955: User.generateTemporaryPassword uses overly-complicated code to generate external URLs
Any references to low-level things like the ``X-Forwarded-Host`` header should be refactored out of User.generateTemporaryPassword and replaced with calls to common utility modules.

### PR DESCRIPTION
**Pull request by zachmullen**
This depends on #984, only the last commit here refers to this topic. I will rebase once 984 goes in, just wanted to put this up.

This contains some code reorganization that also happens to fix #955.

### CODE REVIEW

**Comment by zachmullen** [comment]
Rebased, this is now ready for review.

**Comment by zachmullen** [comment]
@girder/developers anybody want to take a crack at reviewing this?

File: girder/api/v1/user.py [hunk 1]
```diff
@@ -260,28 +260,35 @@ def changeUserPassword(self, user, params):
     def changePassword(self, params):
         self.requireParams(('old', 'new'), params)
         user = self.getCurrentUser()
+        token = None
 
         if not params['old']:
             raise RestException('Old password must not be empty.')
 
         if (not self.model('password').hasPassword(user) or
```
Comment (brianhelba) [inline girder/api/v1/user.py]: If the user has no password, what would we expect ``params['old']`` to contain?

File: girder/api/v1/user.py [hunk 2]
```diff
@@ -312,28 +319,24 @@ def generateTemporaryPassword(self, params):
...
-                'url': url,
-                'token': str(token['_id'])
-            })
-            mail_utils.sendEmail(to=email, subject='Girder: Temporary Access',
-                                 text=html)
+
+        token = self.model('token').createToken(
+            user, days=1, scope=TokenScope.TEMPORARY_USER_AUTH)
+
+        emailHost = self.model('setting').get(SettingKey.EMAIL_HOST)
```
Comment (brianhelba) [inline girder/api/v1/user.py]: Shouldn't you use [``getEmailUrlPrefix``](https://github.com/girder/girder/blob/master/girder/utility/mail_utils.py#L33) here?

File: tests/cases/user_test.py [hunk 3]
```diff
@@ -549,17 +561,9 @@ def testTemporaryPassword(self):
         }, user=user)
         self.assertStatusOk(resp)
 
-        # Generate an email with a forwarded header
```
Comment (brianhelba) [inline tests/cases/user_test.py]: Why aren't you checking whether the email was sent anymore?

File: girder/api/v1/user.py [hunk 4]
```diff
@@ -312,28 +319,24 @@ def generateTemporaryPassword(self, params):
         self.requireParams('email', params)
         email = params['email'].lower().strip()
 
-        users = self.model('user').find({'email': email})
+        user = self.model('user').findOne({'email': email})
 
-        if not users.count():
+        if not user:
             raise RestException('That email is not registered.')
-        for user in users:
```
Comment (brianhelba) [inline girder/api/v1/user.py]: Was there a legitimate reason for this code for the possibility of multiple users with the same email?

Comment (zachmullen) [inline girder/api/v1/user.py]: No, I think the code just predated `findOne`.

File: tests/cases/user_test.py (hunk 3 above)
Comment (zachmullen) [inline tests/cases/user_test.py]: The purpose of creating the email below was to ensure the behavior of the `X-Forwarded-Host` header that was removed as part of this topic branch. That code is obsolete now that we have the email host setting. This is part of the fix for #955

File: girder/api/v1/user.py (hunk 1 above)
Comment (zachmullen) [inline girder/api/v1/user.py]: A user with no password is currently not able to use this endpoint -- the OAuth plugin and any other plugin that allows passwordless users to exist should really hook into this and do something sensible, but as of right now that doesn't exist.

**Comment by brianhelba** [comment]
LGTM.

**Pull request merged**